### PR TITLE
Implement templateTypeRep in DAML

### PR DESCRIPTION
**Pull request by aherrmann-da**
Implements the DAML part of #3072 

Depends on https://github.com/digital-asset/ghc/pull/39
Currently references a testing upload of ghc-lib.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/daml-stdlib-src/DA/Internal/LF.daml [hunk 1]
```diff
@@ -216,3 +217,7 @@ newtype AnyTemplate = AnyTemplate { getAnyTemplate : Any }
 
 -- | Existential choice type that can wrap an arbitrary chice.
 newtype AnyChoice = AnyChoice { getAnyChoice : Any }
+
+-- | Unique textual representation of a template Id.
+newtype TemplateTypeRep = TemplateTypeRep { getTemplateTypeRep : Text }
+  deriving (Eq, Ord)
```
Comment (cocreature) [inline compiler/damlc/daml-stdlib-src/DA/Internal/LF.daml]: We probably want a `MapKey` instance for this which we can easily get via GND but we can leave that for a separate PR.

File: compiler/damlc/tests/daml-test-files/TemplateTypeRep.daml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    wrapped : t
+    p : Party
+  where
+    signatory p
+
+template instance GT1 = GenericT T1
+template instance GT2 = GenericT T2
+
+main = scenario do
+  unless (templateTypeRep @T1 == templateTypeRep @T1) $
```
Comment (cocreature) [inline compiler/damlc/tests/daml-test-files/TemplateTypeRep.daml]: You can simplify these tests using `templateTypeRep @T1 === templateTypeRep @T1`.

Comment (aherrmann-da) [reply compiler/damlc/tests/daml-test-files/TemplateTypeRep.daml]: Not without adding a `Show` instance to `TemplateTypeRep`. I'm not sure if we want that though.

Comment (cocreature) [reply compiler/damlc/tests/daml-test-files/TemplateTypeRep.daml]: Oh I see, yeah I thought about whether we want `Show` or not but haven’t quite made up my mind. I think it’s probably fine to add it given that I wouldn’t expect most users to ever use `templateTypeRep` and for the “expert” users it might be useful for debugging.

Comment (aherrmann-da) [reply compiler/damlc/tests/daml-test-files/TemplateTypeRep.daml]: I've added the `Show` instance and changed the tests to use `===`, I've also added `=/=` for the inequality tests.

File: compiler/damlc/tests/daml-test-files/TemplateTypeRep.daml [hunk 3]
```diff
@@ -0,0 +1,40 @@
+-- Copyright (c) 2019, Digital Asset (Switzerland) GmbH and/or its affiliates.
```
Comment (hurryabit) [inline compiler/damlc/tests/daml-test-files/TemplateTypeRep.daml]: We should also add a test that templates in different packages but with the same name are considered different, i.e.,
```
templateTypeRep @A.T =/= templateTypeRep @B.T
```

Comment (cocreature) [reply compiler/damlc/tests/daml-test-files/TemplateTypeRep.daml]: Do you mean different modules? Of course, they should also be different for different packages but that doesn’t really fit into this test suite and also doesn’t seem to match the example in your comment.

Comment (hurryabit) [reply compiler/damlc/tests/daml-test-files/TemplateTypeRep.daml]: Yes, I meant modules. My bad.
